aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -588,5 +588,208 @@ TORCH_IMPL_FUNC(_convert_indices_from_csr_to_coo_structured_cpu) (
   }
 }
 
+/*
+ * Based on
+ * https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/sparse/sparsetools/csr.h
+ */
+template <class I, class T>
+void _csr_to_block_csr_cpu_kernel(
+    const I n_row,
+    const I n_col,
+    const I R,
+    const I C,
+    const I* input_crow_indices,
+    const I* input_col_indices,
+    const T* input_values,
+    I* result_crow_indices,
+    I* result_col_indices,
+    T* result_values) {
+  // All blocks are possible, that is, may be allocated if a single non-zero
+  // value lives within them. Otherwise they're not.
+
+  // Allocate pointers for all possible column blocks plus 1
+  std::vector<T*> blocks(n_col / C + 1, (T*)0);
+
+  assert(n_row % R == 0);
+  assert(n_col % C == 0);
+
+  // Major assumptions
+  // 1. Blocks must be square
+
+  // Number of blocks along rows
+  I n_brow = n_row / R;
+  // Number of blocks along columns
+  // I n_bcol = n_col / C;
+
+  // Number of elements per block
+  I RC = R * C;
+  // Number of blocks overall
+  I n_blks = 0;
+
+  result_crow_indices[0] = 0;
+
+  // Iterate over blocks along rows
+  for (I block_i = 0; block_i < n_brow; block_i++) {
+    // Iterate over rows within block
+    for (I r = 0; r < R; r++) {
+      I i = R * block_i + r; // row index
+      for (I jj = input_crow_indices[i]; jj < input_crow_indices[i + 1]; jj++) {
+        I j = input_col_indices[jj]; // column index
+
+        // Block corresponding to column index
+        I block_j = j / C;
+        // Column within block
+        I c = j % C;
+
+        if (blocks[block_j] == 0) {
+          blocks[block_j] = result_values + RC * n_blks;
+          result_col_indices[n_blks] = block_j;
+          n_blks++;
+        }
+
+        // Specific blocks entries should not be visited more than once.
+        // Scipy code does an addition here. Why?
```

Comment:
SciPy allows multiple values for the same column like PyTorch allows multiple values for the same coordinate in COO.
We shouldn't allow and encourage that, here's the issue discussing that: https://github.com/pytorch/pytorch/issues/60277.